File: frontend/src/api/bookingApi.js
import axiosClient from './axiosClient';

const bookingApi = {
  /**
   * Kiểm tra phòng còn trống trong khoảng ngày (real-time từ DB).
   *
   * @param {{
   *   roomId?: number | string;
   *   propertySlug?: string;
   *   branchCode?: string;
   *   roomCode?: string;
   *   detailSlug?: string;
   *   checkIn: string;
   *   checkOut: string;
   *   excludeBookingId?: number | string;
   * }} payload
   */
  async checkAvailability(payload) {
    const res = await axiosClient.post('/bookings/check-availability', payload);
    return res.data?.data ?? res.data;
  },

  /**
   * Trạng thái giữ chỗ / đặt phòng theo chi nhánh.
   *
   * @param {{
   *   propertyId?: number | string;
   *   branchId?: number | string;
   *   propertySlug?: string;
   *   branchCode?: string;
   *   from?: string;
   *   to?: string;
   * }} [params]
   */
  async getOccupancy(params = {}) {
    const res = await axiosClient.get('/bookings/occupancy', { params });
    return res.data?.data ?? res.data;
  },

  /**
   * Lịch sử đặt phòng của user đang đăng nhập (có phân trang).
   *
   * @param {{ page?: number; pageSize?: number; filter?: 'all'|'upcoming'|'past'|'pending' }} [params]
   */
  async listMine(params = {}) {
    const res = await axiosClient.get('/bookings/me', { params });
    return res.data?.data ?? res.data;
  },
};

export default bookingApi;

File: mobile/lib/api/booking_api.dart
import 'api_client.dart';

class BookingApi {
  BookingApi(this._client);

  final ApiClient _client;

  Future<Map<String, dynamic>> checkAvailability({
    required String propertySlug,
    required String branchCode,
    required String detailSlug,
    required String checkIn,
    required String checkOut,
  }) async {
    final json = await _client.post('/bookings/check-availability', {
      'propertySlug': propertySlug,
      'branchCode': branchCode,
      'detailSlug': detailSlug,
      'checkIn': checkIn,
      'checkOut': checkOut,
    });
    return json['data'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getOccupancy({
    required String propertySlug,
    required String branchCode,
    String? from,
    String? to,
  }) async {
    final query = <String, String>{
      'propertySlug': propertySlug,
      'branchCode': branchCode,
    };
    if (from != null) query['from'] = from;
    if (to != null) query['to'] = to;
    final json = await _client.get('/bookings/occupancy', query: query);
    return json['data'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> listMine({
    int page = 1,
    int pageSize = 10,
    String filter = 'all',
  }) async {
    final json = await _client.get('/bookings/me', query: {
      'page': '$page',
      'pageSize': '$pageSize',
      'filter': filter,
    });
    return json['data'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> cancelPreview(int bookingId) async {
    final json = await _client.get('/bookings/$bookingId/cancel-preview');
    return json['data'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> cancel(int bookingId) async {
    final json = await _client.post('/bookings/$bookingId/cancel', {});
    return json['data'] as Map<String, dynamic>;
  }
}
